source/agora/node/GossipProtocol.d

```diff
@@ -0,0 +1,97 @@
+/*******************************************************************************
+
+    Contains the code used for peer-to-peer communication.
+
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.node.GossipProtocol;
+import agora.common.Data;
+
+/// Interface of Protocol Owner 
+interface IProtocolOwner {
+@safe:
+
+    /***************************************************************************
+
+        Send a message to all the listeners.
+
+        Params:
+          msg = the message to send
+
+    ***************************************************************************/
+    
+    public void sendMessage(Hash msg); 
+}
+
+/// Procedure of peer-to-peer communication
+class GossipProtocol {
+    private IProtocolOwner owner;
+    private bool[Hash] receivedMsgCache;
+
+    /// Ctor
+    public this (IProtocolOwner owner)
+    {
+        this.owner = owner;
+    }
+
+    /***************************************************************************
+
+        Call this message when receive message.
+
+        Returns:
+            Return true if this message was first received at this node.
+
+        Params:
+          msg = the message received
+
+    ***************************************************************************/
+
+    public bool receiveMessage(Hash msg) @safe
+    {
+        auto p = msg in receivedMsgCache;
+        if (p is null) 
+        {
+            receivedMsgCache[msg] = true;
+            this.owner.sendMessage(msg);
+            return true;
+        } 
+        else
+        {
+            return false;
+        }
+    }
+
+
+    /***************************************************************************
+
+        Check if the message has already been received.
+
+        Returns:
+            Return true if this message was a message already received.
+
+        Params:
+          msg = the message received
+
+    ***************************************************************************/
+
+    public bool hasMessage(Hash msg) @safe
+    {
+        auto p = msg in receivedMsgCache;
+        if (p is null) 
+        {
+            return false;
+        } 
+        else
+        {
+            return true;
+        }
```

Comment:
This whole function can be simplified to:
```
return (msg in receivedMsgCache) !is null;
```

Reply:
@Geod24 Check Please